lib/puppet/state_machine.rb

```diff
@@ -0,0 +1,197 @@
+    #
+    # @example Creating two state machines and composing them in a chain
+    #   machine1 = Puppet::StateMachine.build("Machine 1") do |m|
+    #     m.state(:m1_start,
+    #       action: -> { Puppet.notice("Starting machine 1") },
+    #       event: ->(result) { :started },
+    #       transitions: {started: :m1_final})
+    #
+    #     m.state(:m1_final,
+    #       action: -> { Puppet.notice("Finishing machine 1") },
+    #       type: :final)
+    #   end
+    #
+    #   machine2 = Puppet::StateMachine.build do |m|
+    #     m.state(:m2_start,
+    #       action: -> { Puppet.notice("Starting machine 2") },
+    #       event: ->(result) { :started },
+    #       transitions: {started: :m2_final})
+    #
+    #     m.state(:m2_final,
+    #       action: -> { Puppet.notice("Finishing machine 2") },
+    #       type: :final)
+    #   end
+    #
+    #   composed = Puppet::StateMachine.compose("Composed machine",
+    #   {
+    #     m1: machine1,
+    #     m2: machine2
+    #   }) do |m|
+    #     m.start_state(:composed_start)
+    #
+    #     # Define new start state that transitions to m1
+    #     m.state(:composed_start,
+    #       action: -> { Puppet.notice("Starting composed machine") },
+    #       event: ->(result) { :started },
+    #       transitions: {started: [:m1, :m1_start]})
+    #
+    #     # Define a new final state that m2 will transition to when finished
+    #     m.state(:composed_final,
+    #       type: :final,
+    #       action: -> { Puppet.notice("Finishing composed machine") })
+    #
+    #     # Transition from m1 when finished to m2 start
+    #     m.transition(
+    #       source: [:m1, :m1_final],
+    #       target: [:m2, :m2_start],
+    #       on: Puppet::StateMachine::FINAL)
+    #
+    #     # Transition from m2 final state to composed machine final state
+    #     m.transition(
+    #       source: [:m2, :m2_final],
+    #       target: :composed_final,
+    #       on: Puppet::StateMachine::FINAL)
+    #
+    #   end
+    #
+    def self.compose(name, machines, &block)
+      MachineBuilder.new(name).compose(machines, &block)
+    end
+
+    # @!attribute [r] name
+    #   @return [String] The description of this state machine
+    attr_reader :name
+
+    # @!attribute [r] states
+    #   @api private
+    attr_reader :states
+
+    # @param name [String] The description of this state machine.
+    # @param states [Hash<Symbol, Puppet::StateMachine::State>]
+    # @param start_state [Symbol] The name of the initial state used when executing this machine.
+    def initialize(name, states, start_state)
+      @name = name
+      @states = states
+      @start_state = start_state
+    end
+
+    # @return [Puppet::StateMachine::State]
+    def start_state
+      @states[@start_state]
+    end
+
+    def state(state_name)
+      @states[state_name]
+    end
+
+    # Generate a set of state machines with the given namespace added to the state name
+    # and transitions.
```

Comment:
Is this comment right? It seems like it generates a set of machine states, rather than state machines.